### PR TITLE
When setting suggestion for significant_drop_in_scrutinee, add suggestion for MoveAndClone for non-ref

### PR DESCRIPTION
**Pull request by PrestonFrom**
When trying to set the current suggestion, if the type of the expression
is not a reference and it is not trivially pure clone copy, we should still
trigger and emit a lint message. Since this fix may require cloning an
expensive-to-clone type, do not attempt to offer a suggested fix.

This change means that matches generated from TryDesugar and AwaitDesugar
would normally trigger a lint, but they are out of scope for this lint,
so we will explicitly ignore matches with sources of TryDesugar or
AwaitDesugar.

changelog: Update for ``[`significant_drop_in_scrutinee`]`` to correctly
emit lint messages for cases where the type is not a reference *and*
not trivially pure clone copy.
changelog: [`significant_drop_in_scrutinee`]: No longer lint on Try `?` 
and `await` desugared expressions.

### CODE REVIEW

**Comment by matthiaskrgr** [comment]
what is this 4 mb "main" file? 🤔
is this intended to go into the repo?

**Comment by PrestonFrom** [comment]
@matthiaskrgr That is a massive oversight on my part! It should not be there. Deleting right away!

**Review by flip1995 — COMMENTED** [review]
Changes LGTM. One comment on the lint message for `for` loops.

File: tests/ui/significant_drop_in_scrutinee.stderr [hunk 1]
```diff
@@ -279,5 +279,17 @@ LL ~     i += mutex.lock().unwrap().i;
...
    |
 
-error: aborting due to 23 previous errors
+error: temporary with significant drop in match scrutinee
+  --> $DIR/significant_drop_in_scrutinee.rs:579:11
+   |
+LL |     match rwlock.read().unwrap().to_number() {
+   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: temporary with significant drop in match scrutinee
```
[the remark this answers is not in the document]
Comment (PrestonFrom) [reply tests/ui/significant_drop_in_scrutinee.stderr]: That is a great point! Updated to correctly identify if it's in a match or a for loop.

**Comment by PrestonFrom** [comment]
@flip1995 Thank you for the review! I pushed the change as a new commit -- assuming it looks good, I should push again with it squashed, right?

**Comment by flip1995** [comment]
No need to squash here. The second commit is also a functional change as opposed to formatting or typo fixes or "addressing review" commits. I'm good with keeping them around, because it also shows in the history that there was work dedicated to getting this right.

Thanks for addressing this so quickly!

@bors r+

**Pull request merged**